docs/design/ovs-pipeline.md

```diff
@@ -1,1197 +1,1938 @@
-          app: nginx
-    ports:
+    - from:
+        - podSelector:
+            matchLabels:
+              app: client
+      ports:
+        - protocol: TCP
+          port: 80
+  egress:
+    - to:
+        - podSelector:
+            matchLabels:
+              app: db
+      ports:
+        - protocol: TCP
+          port: 3306
+```
+
+## Kubernetes Service Implementation
+
+Like Kubernetes NetworkPolicy, several tables of the pipeline are dedicated to [Kubernetes
+Service](https://kubernetes.io/docs/concepts/services-networking/service/) implementation (table [NodePortMark],
+[SessionAffinity], [ServiceLB], and [EndpointDNAT]).
+
+By enabling `proxyAll`, ClusterIP, NodePort, LoadBalancer, and ExternalIP are all supported. Otherwise, only in-cluster
+ClusterIP is supported. In this document, we use the sample Kubernetes Services below. These Services select Pods with
+the label `app: web` as Endpoints.
+
+### ClusterIP without Endpoint
+
+A sample Service with `clusterIP` set to `10.101.255.29` does not have any associated Endpoint.
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: sample-clusterip-no-ep
+spec:
+  ports:
     - protocol: TCP
       port: 80
-  egress:
-  - to:
-    - podSelector:
-        matchLabels:
-          app: nginx
-    ports:
+      targetPort: 80
+  clusterIP: 10.101.255.29
+```
+
+### ClusterIP
+
+A sample ClusterIP Service with `clusterIP` set to `10.105.31.235`.
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: sample-clusterip
+spec:
+  selector:
+    app: web
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 80
+  clusterIP: 10.105.31.235
+```
+
+### ClusterIP with Session Affinity
+
+A sample ClusterIP Service with `clusterIP` set to `10.96.76.15` configured with session affinity.
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: sample-clusterip-session-affinity
+spec:
+  selector:
+    app: web
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 80
+  clusterIP: 10.96.76.15
+  sessionAffinity: ClientIP
+  sessionAffinityConfig:
+    clientIP:
+      timeoutSeconds: 300
+```
+
+### NodePort
+
+A sample NodePort Service with a `nodePort` set to `30004`.
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: sample-nodeport
+spec:
+  selector:
+    app: web
+  ports:
     - protocol: TCP
       port: 80
+      targetPort: 80
+      nodePort: 30004
+  type: NodePort
 ```
 
-This Network Policy is applied to all Pods with the `nginx` app label in the
-`default` Namespace. For these Pods, it only allows TCP traffic on port 80 from
-and to Pods which also have the `nginx` app label. Because Antrea will only
-install OVS flows for this Network Policy on Nodes for which some of the Pods
-are the target of the policy, we have scheduled 2 `nginx` Pods on the same
-Node. They received IP addresses 10.10.1.2 and 10.10.1.3 from the Antrea CNI, so
-you will see these addresses show up in the OVS flows.
+### NodePort with Session Affinity
```

Comment:
do you think we really need all the "Session Affinity" variants? I think "ClusterIP with Session Affinity" should be enough. If there is a significant difference in the implementation between "internal" traffic and "external" traffic, let me know. But even if there is, maybe we could keep a single "external" Service with Session Affinity.

Reply:
We don't need all "Session Affinity" variants. It is just an attribute of Service, and the Service could be ClusterIP, NodePort, or LoadBalancer. There is no significant difference in our implementation of "Session Affinity" for different types of Services. I use a general Service to demonstrate "Session Affinity" in the latest commit.